File: common/limiter/lua/sliding_window.lua
-- Atomic sliding-window rate limiter.
--
-- The equivalent Go code used to issue LLEN, then LINDEX, then LPUSH as separate
-- round trips. Between the read and the write, every concurrent caller observed
-- the same under-limit length and every one of them was admitted, so the limiter
-- only ever constrained sequential traffic. Running the whole decision inside
-- Redis makes it atomic.
--
-- KEYS[1] bucket key
-- ARGV[1] maximum requests allowed inside the window
-- ARGV[2] window length in seconds
-- ARGV[3] current unix time in seconds
-- ARGV[4] key ttl in seconds
--
-- Returns 1 when the request is admitted, 0 when it is rejected.

local key    = KEYS[1]
local maxNum = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local now    = tonumber(ARGV[3])
local ttl    = tonumber(ARGV[4])

if maxNum <= 0 then
  return 0
end

local length = redis.call('LLEN', key)
if length < maxNum then
  redis.call('LPUSH', key, now)
  redis.call('EXPIRE', key, ttl)
  return 1
end

-- Entries written by the previous implementation are formatted timestamps rather
-- than numbers. tonumber yields nil for those, and treating that as "window has
-- rolled" lets the key heal itself within one window instead of erroring.
local oldest = tonumber(redis.call('LINDEX', key, -1))
if oldest ~= nil and (now - oldest) < window then
  redis.call('EXPIRE', key, ttl)
  return 0
end

redis.call('LPUSH', key, now)
redis.call('LTRIM', key, 0, maxNum - 1)
redis.call('EXPIRE', key, ttl)
return 1

File: common/limiter/sliding_window.go
package limiter

import (
	"context"
	_ "embed"
	"time"

	"github.com/go-redis/redis/v8"
)

//go:embed lua/sliding_window.lua
var slidingWindowScript string

// slidingWindow is safe for concurrent use: redis.Script caches the SHA and
// falls back to a plain EVAL when Redis reports NOSCRIPT, so a Redis restart
// that clears the script cache repairs itself instead of failing every request.
var slidingWindow = redis.NewScript(slidingWindowScript)

// AllowSlidingWindow reports whether one more request fits inside the window for
// key. The check and the record happen in a single atomic script; splitting them
// across round trips lets a concurrent burst through the limit entirely.
func AllowSlidingWindow(ctx context.Context, rdb *redis.Client, key string, maxRequestNum int, window time.Duration, ttl time.Duration) (bool, error) {
	result, err := slidingWindow.Run(ctx, rdb, []string{key},
		maxRequestNum,
		int64(window.Seconds()),
		time.Now().Unix(),
		int64(ttl.Seconds()),
	).Int()
	if err != nil {
		return false, err
	}
	return result == 1, nil
}
